### PR TITLE
Adds colocalization analysis and refactors code to make it similar to GWAS UX

### PR DESCRIPTION
**Pull request by mcamarad**
# Summary:

It allows running colocalization analysis with summary statistics from a GWAS computed with SAIGE:
- Uses output from `lifebit-ai/gel-gwas` to test whether the trait of interest has common variants with the different phenotypes resulting from a pheWAS.

It refactors and updates the nextflow variables & `transform_cb_output.R` to support continuous covariates

# Tasks:

- **(1) Make it compatible with SAIGE GWAS output**
    - [x] Check which stats are needed.
    - [x] Check that format of SNPs is compatible
    - [x] Check that `coloc.abf()` runs with the data provided from SAIGE
- **(2) Build script to run quantitative and binary coloc analysis**
    - [x] Create function to run across all phenotypes.
    - [x] Produce final table & plot
- **(3) Modify report to include colocalization analysis when selected**
    - [x] Add new table
    - [x] Add new plot
- **(4) Add necessary nextflow processes**
    - [x] Create switch for turning the analysis on and off 
    - [x] Bifurcate report into two process depending on scenarios
- **(6) Update nextflow.config and Dockerfile**
- **(7) Adds transformation with continuous covariates**
- **(8) adapts pipeline usage to be similar to `gel-gwas`**

# Comments:
After testing with GWAS catalogue summary stats, some information was missing and it was not possible yet to use it within this pipeline. Main issue:
- Lack of Sample size data (Total, case, controls) which is essential for `coloc` analysis.

# Test in Lifebit CloudOS
Latest test completed: https://cloudos.lifebit.ai/app/jobs/5f878fc28f81710113099d68

### CODE REVIEW

File: bin/run_coloc.R [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+colnames(coloc_results) = c("n_snps", "pp_H0", "pp_H1", "pp_H2", "pp_H3", "pp_H4")
+
+########################################
+### Plot heatmap with results
+########################################
+
+group_col_fun = structure(brewer.pal(length(grouping), "Set3"), 
+    names = grouping)
+ha = rowAnnotation(group = anno_simple(grouping),
```
Comment (cgpu) [inline bin/run_coloc.R:104]: The `package::function()` notation leaves no room for doubt, "which packages is this function exported from"?

```Suggestion
ha = ComplexHeatmap::rowAnnotation(group = anno_simple(grouping), 
```

File: bin/run_coloc.R [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+### Plot heatmap with results
+########################################
+
+group_col_fun = structure(brewer.pal(length(grouping), "Set3"), 
+    names = grouping)
+ha = rowAnnotation(group = anno_simple(grouping), 
+                   col = group_col_fun,
+                   annotation_name_side = "bottom",
+                   show_legend = TRUE)
+ht = Heatmap(coloc_results[,-1],
```
Comment (cgpu) [inline bin/run_coloc.R:108]: The `package::function()` notation leaves no room for doubt, "which packages is this function exported from"?

```Suggestion
ht = ComplexHeatmap::Heatmap(coloc_results[,-1],
```

File: bin/phewas_report.Rmd [hunk 3]
```diff
@@ -36,30 +38,77 @@ library(DT)
...
+table   <- as.data.frame(data.table::fread(params$phewas_results))
 colnames(table) <- snakecase::to_snake_case(colnames(table))
 DTable(table)
 ```
+
+# Post Analysis
+
+## Colocalization
+
+The following analysis performs genetic colocalization analysis of two potentially related phenotypes, to ask whether they share common genetic causal variant(s). The hypothesis tested are the following:
```
Comment (cgpu) [inline bin/phewas_report.Rmd:74]: Minor, plural needed.

```Suggestion
The following analysis performs genetic colocalization analysis of two potentially related phenotypes, to ask whether they share common genetic causal variant(s). The hypotheses tested are the following:
```

File: bin/phewas_report.Rmd [hunk 4]
```diff
@@ -36,30 +38,77 @@ library(DT)
...
+
+- H0: neither trait has a genetic association in the region
+- H1: only PheWAS phenotype has a genetic association in the region
+- H2: only GWAS trait has a genetic association in the region
+- H3: both traits are associated, but with different causal variants
+- H4: both traits are associated and share a single causal variant
+
+The plot below will display the Posterior Probability (PP) associated to all different hypothesis.
```
Comment (cgpu) [inline bin/phewas_report.Rmd:82]: Minor, plural needed.

```Suggestion
The plot below will display the Posterior Probability (PP) associated to all different hypotheses.
```

File: bin/run_coloc.R [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+### Prepare to plot
+########################################
+
+grouping = phewas_df %>% select(description, group) %>% distinct() %>% arrange(description) %>% select(group)
+grouping = grouping$group
+
+colnames(coloc_results) = c("n_snps", "pp_H0", "pp_H1", "pp_H2", "pp_H3", "pp_H4")
+
+########################################
+### Plot heatmap with results
```
Comment (cgpu) [inline bin/run_coloc.R:99]: Aesthetic comment, the colors are a bit out-of-the box and this makes the report look a bit less coherent.

Ideally we implement always with custom colors to be able to swap hex codes at will. 

I checked a bit, I think what we might need is the following:

```r
library(circlize)
col_fun = colorRamp2(c(-2, 0, 2), c("#4dc5ce", "#92C581", "#5c5c5e"))
ComplexHeatmap::Heatmap(mat, name = "mat", col = col_fun)
```

source: https://jokergoo.github.io/ComplexHeatmap-reference/book/a-single-heatmap.html#colors

File: docs/cb_integration_phewas.md [hunk 6]
```diff
@@ -128,6 +128,18 @@ rs10494464
 https://cloudos.lifebit.ai/app/jobs/5f7fc50a8f81710113038a46 :tada:
 
 ## 5. Future Work
+<<<<<<< HEAD
```
Comment (cgpu) [inline docs/cb_integration_phewas.md:131]: Merge conflict sneaked in a bit, can you check which one we want to keep?

File: docs/cb_integration_phewas.md [hunk 7]
```diff
@@ -137,6 +149,7 @@ https://cloudos.lifebit.ai/app/jobs/5f7fc50a8f81710113038a46 :tada:
     - [ ] Refactor script so when aggregated VCFs & CB inputs are given it run subsequent processes, similar to GWAS
     - [ ] Add Prepare aggregate VCF step
     - [ ] Add CB output integration
+>>>>>>> master
```
Comment (cgpu) [inline docs/cb_integration_phewas.md:152]:
```Suggestion
```

File: docs/cb_integration_phewas.md [hunk 8]
```diff
@@ -128,6 +128,18 @@ rs10494464
...
+- **(2) Fix multiple contrast design and branching of processes**
+- **(3) Add transformations for continuous variables**
+- **(4) Refactor round**
+    - [ ] Make sure everything looks tidy and params looks consistent
+- **(5) prepare nextflow processes for integration with VCFs**
+    - [ ] Refactor script so when aggregated VCFs & CB inputs are given it run subsequent processes, similar to GWAS
+    - [ ] Add Prepare aggregate VCF step
+    - [ ] Add CB output integration
+- **(6) Add support to continuous variables as PHE**
+=======
```
Comment (cgpu) [inline docs/cb_integration_phewas.md:142]: Here you need to resolve the conflict and choose which text to keep.

File: docs/colocalization.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+### Iter2
+
+https://cloudos.lifebit.ai/app/jobs/5f878fc28f81710113099d68 :tada:
+
+## 5. Future Work
+- **(1) Add GWAS catalogue**
+- **(2) Improve plot, and explore other plotting options**
+- **(3) Add other types of data for colocalization (eQTLs, pQTLs)**
+- **(4) Add scatterplots for comparison between genetic variants of interest between datasets**
```
Comment (cgpu) [inline docs/colocalization.md:62]: The above are probably not needed info for the end user, we can safely migrate in https://github.com/lifebit-ai/gwas-utils/tree/master/docs/phewas

File: docs/usage_and_parameters.md [hunk 10]
```diff
@@ -1,39 +1,44 @@
...
-- **continuous_var_transformation** : Transforms continuous variables using 'log', 'log10', 'log2', 'zscores' or 'None'.
-- **continuous_var_aggregation** : Defines how to aggregate different measurements. Choose between 'max', 'min', 'median' or 'mean'.
-- **snps** : File containing list of SNPs to be tested.
-- **snp_threshold** : Threshold to select SNPs significance from plink association analysis if `snps` are not provided. Defaults to 0.05
-- **outdir** : Output directory for results.
-- **output_tag** : Prefix to identify output files.
+- **--continuous_var_transformation** : Transforms continuous variables using 'log', 'log10', 'log2', 'zscores' or 'None'.
+- **--continuous_var_aggregation** : Defines how to aggregate different measurements. Choose between 'max', 'min', 'median' or 'mean'.
+- **--snps** : File containing list of SNPs to be tested.
+- **--snp_threshold** : Threshold to select SNPs significance from plink association analysis if `snps` are not provided. Defaults to 0.05
```
Comment (cgpu) [inline docs/usage_and_parameters.md:33]: What metric is the threshold for, p-value? If so might be worth making a bit more explict.

```Suggestion
- **--snp_p_val_threshold** : Threshold to select SNPs significance from plink association analysis if `snps` are not provided. Defaults to 0.05

```

File: main.nf [hunk 11]
```diff
@@ -432,58 +428,110 @@ if (params.plink_input && params.phenofile && params.metadata) {
...
-// process filter {
-//     publishDir "${params.outdir}/filter", mode: 'copy'
+        output:
+        set file("*coloc_heatmap.png"), file("*coloc_results.csv") into coloc_results_ch
 
-//     input:
-//     set file(bed), file(bim), file(fam) from plink
+        script:
+        """
+        cp /opt/bin/* .
```
Comment (cgpu) [inline main.nf:449]: No need since we will have it in Dockerfile.

```Suggestion
```

File: main.nf [hunk 12]
```diff
@@ -243,20 +246,19 @@ if (params.plink_input && params.phenofile && params.metadata) {
     script:
     """
     cp /opt/bin/* .
```
Comment (cgpu) [inline main.nf:248]: We have the files in PATH, no need to copy contents of bin as it will be accessible.

```Suggestion
```

File: main.nf [hunk 13]
```diff
@@ -273,20 +275,18 @@ if (params.plink_input && params.phenofile && params.metadata) {
         """
         cp /opt/bin/* .
```
Comment (cgpu) [inline main.nf:276]: We have the files in PATH, no need to copy contents of bin as it will be accessible.

```Suggestion
```

File: main.nf [hunk 14]
```diff
@@ -302,19 +302,17 @@ if (params.plink_input && params.phenofile && params.metadata) {
             """
             cp /opt/bin/* .
```
Comment (cgpu) [inline main.nf:303]: We have the files in PATH, no need to copy contents of bin as it will be accessible.

```Suggestion
```

File: main.nf [hunk 15]
```diff
@@ -325,19 +323,17 @@ if (params.plink_input && params.phenofile && params.metadata) {
             file(json) from ch_encoding_json
 
             output:
-            file("${output_tag}_design_matrix_control_*.phe'") into pheno, pheno2
+            file("${output_tag}_design_matrix_control_*.phe'") into (pheno, pheno2)
 
             script:
             """
             cp /opt/bin/* .
```
Comment (cgpu) [inline main.nf:330]: We have the files in PATH, no need to copy contents of bin as it will be accessible.

```Suggestion
```

**Review by cgpu — CHANGES_REQUESTED** [review]
Some minor changes requested, 

- `cp /opt/bin/* .`	is not needed when using the script, exception an Rmd file if it is looking for R scripts from within
- We can make the reports look neater by adapting hex codes

**Comment by mcamarad** [comment]
@cgpu Thanks a lot for the great comments!!! I added the changes and this is the job in CloudOS:

https://cloudos.lifebit.ai/app/jobs/5f89ade78f817101130ca444

**Review by cgpu — APPROVED** [review]
Ready, steady, squash! 🏁 🏎️

**Pull request merged**